src/v/cloud_topics/batcher/aggregator.cc

```diff
@@ -0,0 +1,190 @@
+          b.base);
+
+        // TX data (producer id, control flag) are not copied from 'src' yet.
+
+        // Put the payload
+        builder.add_raw_kv(
+          serde::to_iobuf(dl_placeholder_record_key::payload),
+          serde::to_iobuf(placeholder));
+
+        // TODO: fix this
+        for (int i = 1; i < b.num_records - 1; i++) {
+            iobuf empty;
+            builder.add_raw_kv(
+              serde::to_iobuf(dl_placeholder_record_key::empty),
+              std::move(empty));
+        }
+        result->placeholders.push_back(std::move(builder).build());
+        ctx.size_bytes += b.size_bytes;
+    }
+    result->ref = req.weak_from_this();
+    ctx.placeholders.push_back(std::move(result));
+}
+} // namespace
+
+template<class Clock>
+chunked_vector<std::unique_ptr<batches_for_req<Clock>>>
+aggregator<Clock>::get_placeholders() {
+    prepared_placeholder_batches<Clock> ctx{
+      .id = _id,
+    };
+    for (auto& [key, list] : _staging) {
+        for (auto& req : list) {
+            vassert(
+              !req.data_chunk.payload.empty(),
+              "Empty write request for ntp: {}",
+              key);
+            make_dl_placeholder_batches(ctx, req, req.data_chunk);
+        }
+    }
+    return std::move(ctx.placeholders);
+}
+
+template<class Clock>
+iobuf aggregator<Clock>::get_stream() {
+    iobuf concat;
+    for (auto& p : _aggregated) {
+        if (p->ref != nullptr) {
+            concat.append(std::move(p->ref->data_chunk.payload));
+        }
+    }
+    return concat;
+}
+
+template<class Clock>
+object_id aggregator<Clock>::get_object_id() const noexcept {
+    return _id;
+}
+
+template<class Clock>
+iobuf aggregator<Clock>::prepare() {
+    // Move data from staging to aggregated
+    _aggregated = get_placeholders();
+    _staging.clear();
+    // Produce input stream
+    return get_stream();
+}
+
+template<class Clock>
+void aggregator<Clock>::ack() {
+    if (_aggregated.empty()) {
+        return;
+    }
+    auto d = ss::defer([this] { _aggregated.clear(); });
+    for (auto& p : _aggregated) {
+        if (p->ref != nullptr) {
+            try {
+                p->ref->set_value(std::move(p->placeholders));
+            } catch (const ss::broken_promise&) {
+            }
+        }
+    }
+}
+
+template<class Clock>
+void aggregator<Clock>::ack_error(errc e) {
+    if (_aggregated.empty()) {
+        return;
+    }
+    auto d = ss::defer([this] { _aggregated.clear(); });
+    for (auto& p : _aggregated) {
+        if (p->ref != nullptr) {
+            try {
+                p->ref->set_value(e);
+            } catch (const ss::broken_promise&) {
+            }
+        }
+    }
+}
+
+template<class Clock>
+void aggregator<Clock>::add(write_request<Clock>& req) {
+    auto it = _staging.find(req.ntp);
+    if (it == _staging.end()) {
+        _staging.insert(std::make_pair(req.ntp, write_request_list<Clock>()));
+        it = _staging.find(req.ntp);
```

Comment:
You can use returned iterator from `insert` or `emplace` to avoid this extra `find()`, and returned `bool` for `vassert()`